### PR TITLE
[HWORKS-225] Hopsworks ha

### PR DESCRIPTION
**Pull request by ErmiasG**

### CODE REVIEW

File: providers/configure_server.rb [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    realm_name "#{realmname}"
+    jaas_context "jdbcRealm"
+    properties props
+    domain_name domain_name
+    password_file password_file
+    username username
+    admin_port admin_port
+    secure false
+    classname "com.sun.enterprise.security.auth.realm.jdbc.JDBCRealm"
+    not_if "#{asadmin_cmd} list-auth-realms #{target} | grep kthfsrealm"
```
Comment (kouzant) [inline providers/configure_server.rb:226]: Replace the hardcoded realm with the variable

File: recipes/das_node.rb [hunk 2]
```diff
@@ -0,0 +1,485 @@
+#rows_path = "#{domains_dir}/post.sql"
+
+case node['platform']
+when "rhel"
+  package "openssh-clients"
```
Comment (kouzant) [inline recipes/das_node.rb:5]: Do we still need this?

File: recipes/das_node.rb [hunk 3]
```diff
@@ -0,0 +1,485 @@
...
+end
+
+package "openssh-server"
```
Comment (kouzant) [inline recipes/das_node.rb:8]: And this?

File: recipes/das_node.rb [hunk 4]
```diff
@@ -0,0 +1,485 @@
...
+public_ip=my_public_ip()
```
Comment (kouzant) [inline recipes/das_node.rb:9]: Use `my_private_ip` instead

File: recipes/das_node.rb [hunk 5]
```diff
@@ -0,0 +1,485 @@
...
+admin_port = node['hopsworks']['admin']['port']
+username=node['hopsworks']['admin']['user']
+password=node['hopsworks']['admin']['password']
+
+config_nodes= node['hopsworks'].attribute?('config_node')? private_recipe_ips('hopsworks', 'config_node') : []
+
+current_version = node['hopsworks']['current_version']
+
+asadmin = "#{node['glassfish']['base_dir']}/versions/current/bin/asadmin"
+password_file = "#{domains_dir}/#{domain_name}_admin_passwd"
```
Comment (kouzant) [inline recipes/das_node.rb:26]: If it's not too bothersome can you create helper functions to construct them? I've seen them in many places

File: recipes/das_node.rb [hunk 6]
```diff
@@ -0,0 +1,485 @@
...
+  admin_port admin_port
+  target payara_config
+  asadmin asadmin
+  internal_port node['hopsworks']['internal']['port']
+  network_name "https-internal"
+  network_listener_name "https-int-list"
+  action :glassfish_configure_network
+end
+
+if node['hopsworks']['ha']['loadbalancer'].to_s == "true"
```
Comment (kouzant) [inline recipes/das_node.rb:115]: Use `casecmp?` for comparison which is case-insensitive

File: recipes/das_node.rb [hunk 7]
```diff
@@ -0,0 +1,485 @@
...
+  mode '0750'
+end
+
+directory node['hopsworks']['data_volume']['node_logs'] do
+  owner node['glassfish']['user']
+  group node['glassfish']['group']
+  mode '0750'
+end
+
+bash 'Move glassfish logs to data volume' do
```
Comment (kouzant) [inline recipes/das_node.rb:205]: This is not needed I think. The purpose of this was to move old directories to the data volume. If you already create the directory in the data volume and symlink it then this snippet is not needed.

File: recipes/config_node.rb [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+asadmin = "#{node['glassfish']['base_dir']}/versions/current/bin/asadmin"
+password_file = "#{domains_dir}/#{domain_name}_admin_passwd"
+username=node['hopsworks']['admin']['user']
+nodedir=node['glassfish']['nodes_dir']
+payara_config = "hopsworks-config"
+deployment_group = "hopsworks-dg"
+
+admin_port = node['hopsworks']['admin']['port']
+das_ip=private_recipe_ip('hopsworks', 'das_node')
+public_ip=my_public_ip()
```
Comment (kouzant) [inline recipes/config_node.rb:12]: Replace with `my_private_ip`

File: providers/worker.rb [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+              restart_domain_command: "#{restart_instance_command}",
+              stop_domain_command: "#{stop_instance_command}",
+              start_domain_timeout: systemd_start_timeout,
+              stop_domain_timeout: systemd_stop_timeout)
+    notifies :start, "service[#{service_name}]", :delayed
+  end
+
+  service service_name do
+    supports start: true, restart: true, stop: true, status: true
+    action [:enable]
```
Comment (kouzant) [inline providers/worker.rb:44]: Add a guard here if `node['services']['enabled']` is "true"

**Review by kouzant — CHANGES_REQUESTED** [review]

File: recipes/das_node.rb (hunk 5 above)
Comment (ErmiasG) [reply recipes/das_node.rb:18]: I thought about it too but I do not think it will be that useful.

**Review by kouzant — CHANGES_REQUESTED** [review]

File: recipes/install.rb [hunk 10]
```diff
@@ -465,7 +525,7 @@
 directory node['hopsworks']['domain1']['logs'] do
   recursive true
   action :delete
-  not_if { conda_helpers.is_upgrade }
+  only_if { Dir.entries(node['hopsworks']['domain1']['logs']).select {|entry| !(entry =='.' || entry == '..' || entry == '.gitkeep') }.empty? || !conda_helpers.is_upgrade }
```
[the remark this answers is not in the document]
Comment (ErmiasG) [reply recipes/install.rb:529]: The dir is not empty.

**Review by kouzant — APPROVED** [review]

**Pull request merged**